pkg/operators/wasm/perf.go

```diff
@@ -0,0 +1,242 @@
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "perfReaderSetDeadline", i.perfReaderSetDeadline,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // PerfReader
+			wapi.ValueTypeI64, // Nanoseconds
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "perfReaderRead", i.perfReaderRead,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // PerfReader
+			wapi.ValueTypeI32, // Buf pointer address
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "perfReaderClose", i.perfReaderClose,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // PerfReader
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+}
+
+// newPerfReader creates a new perf reader.
+// Params:
+// - stack[0] is the perf map handle
+// - stack[1] is the size
+// - stack[2] is a boolean indicating whether the perf buffer is overwritable
+// - stack[3] is the value size
+// - stack[4] is the max entries
+// Return value:
+// - Perf reader handle on success, 0 on error
+func (i *wasmOperatorInstance) newPerfReader(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfMapHandle := wapi.DecodeU32(stack[0])
+	size := int(wapi.DecodeU32(stack[1]))
+	isOverwritable := wapi.DecodeU32(stack[2]) > 0
+
+	perfMap, ok := getHandle[*ebpf.Map](i, perfMapHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	perfReader, err := perf.NewReaderWithOptions(perfMap, size, perf.ReaderOptions{Overwritable: isOverwritable})
+	if err != nil {
+		i.logger.Warnf("newPerfReader: creating perf reader: %v", err)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = wapi.EncodeU32(i.addHandle(perfReader))
+}
+
+// perfReaderPause() pauses the perf reader
+// Params:
+// - stack[0]: Perf reader handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) perfReaderPause(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	err := perfReader.Pause()
+	if err != nil {
+		i.logger.Warnf("perfReaderPause: pausing perf reader: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	stack[0] = 0
+}
+
+// perfReaderResume() resume the perf reader
+// Params:
+// - stack[0]: Perf reader handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) perfReaderResume(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	err := perfReader.Resume()
+	if err != nil {
+		i.logger.Warnf("perfReaderResume: resuming perf reader: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	stack[0] = 0
+}
+
+// perfReaderSetDeadline() set the perf reader deadline
+// Params:
+// - stack[0]: Perf reader handle
+// - stack[1]: Time in nanoseconds
+// Return value:
+// - 0 on success, 1 on error
```

Comment:
Do we need to expose that to the wasm module?

AFAIK, so far, the functions in the wasm module don't sleep. With `perfReaderRead` and `perfReaderSetDeadline`, you are changing that. This has the potential to introduce tricky bugs. For example, if a gadget sets up a long deadline, and the user is trying to stop the gadget. If the wasm runner is blocked, IG would not be able to stop the gadget.

It seems to me that the event loop should be implemented in IG core instead of in the wasm guest. And for traceloop, we don't need blocking reads anyway, so it seems we are introducing complex API without a use case for it.